File: bot-api/java/src/test/java/test_utils/TestBotBuilder.java
package test_utils;

import dev.robocode.tankroyale.botapi.BaseBot;
import dev.robocode.tankroyale.botapi.Bot;
import dev.robocode.tankroyale.botapi.BotInfo;
import dev.robocode.tankroyale.botapi.InitialPosition;
import dev.robocode.tankroyale.botapi.events.*;

import java.net.URI;
import java.util.function.Consumer;

/**
 * Builder/factory for creating configurable test bots.
 * <p>
 * This builder allows tests to create bots with specific behaviors without
 * creating separate bot classes for each test scenario. It supports:
 * <ul>
 *   <li>Predefined behavior profiles (passive, aggressive, scanning)</li>
 *   <li>Custom event handler callbacks for targeted testing</li>
 *   <li>Fluent API for easy configuration</li>
 * </ul>
 * <p>
 * Example usage:
 * <pre>{@code
 * var bot = TestBotBuilder.create()
 *     .withBehavior(BotBehavior.PASSIVE)
 *     .onTick(e -> bot.setTurnRate(5))
 *     .build(serverUrl);
 * }</pre>
 */
public class TestBotBuilder {

    /**
     * Predefined bot behavior profiles.
     */
    public enum BotBehavior {
        /**
         * Does nothing - just calls go() each tick.
         */
        PASSIVE,

        /**
         * Moves forward and fires continuously.
         */
        AGGRESSIVE,

        /**
         * Continuously rotates radar to scan for enemies.
         */
        SCANNING,

        /**
         * Custom behavior defined by callbacks only.
         */
        CUSTOM
    }

    // Bot info configuration
    private String name = "TestBot";
    private String version = "1.0";
    private String[] authors = {"Test Author"};
    private String description = "Test bot for unit tests";
    private String platform = "JVM";
    private String programmingLang = "Java";

    // Behavior configuration
    private BotBehavior behavior = BotBehavior.PASSIVE;

    // Event callbacks
    private Consumer<TickEvent> onTickCallback;
    private Consumer<ScannedBotEvent> onScannedBotCallback;
    private Consumer<HitBotEvent> onHitBotCallback;
    private Consumer<HitWallEvent> onHitWallCallback;
    private Consumer<BulletFiredEvent> onBulletFiredCallback;
    private Consumer<HitByBulletEvent> onHitByBulletCallback;
    private Consumer<BulletHitBotEvent> onBulletHitBotCallback;
    private Consumer<DeathEvent> onDeathCallback;
    private Consumer<RoundStartedEvent> onRoundStartedCallback;
    private Consumer<RoundEndedEvent> onRoundEndedCallback;
    private Consumer<GameStartedEvent> onGameStartedCallback;
    private Consumer<GameEndedEvent> onGameEndedCallback;
    private Runnable onRunCallback;

    private TestBotBuilder() {
    }

    /**
     * Create a new TestBotBuilder instance.
     *
     * @return a new builder
     */
    public static TestBotBuilder create() {
        return new TestBotBuilder();
    }

    /**
     * Set the bot name.
     *
     * @param name the bot name
     * @return this builder
     */
    public TestBotBuilder withName(String name) {
        this.name = name;
        return this;
    }

    /**
     * Set the bot version.
     *
     * @param version the bot version
     * @return this builder
     */
    public TestBotBuilder withVersion(String version) {
        this.version = version;
        return this;
    }

    /**
     * Set the bot authors.
     *
     * @param authors the bot authors
     * @return this builder
     */
    public TestBotBuilder withAuthors(String... authors) {
        this.authors = authors;
        return this;
    }

    /**
     * Set the bot behavior profile.
     *
     * @param behavior the predefined behavior
     * @return this builder
     */
    public TestBotBuilder withBehavior(BotBehavior behavior) {
        this.behavior = behavior;
        return this;
    }

    /**
     * Set a callback for the onTick event.
     * This callback is invoked AFTER any behavior-specific tick handling.
     *
     * @param callback the tick event callback
     * @return this builder
     */
    public TestBotBuilder onTick(Consumer<TickEvent> callback) {
        this.onTickCallback = callback;
        return this;
    }

    /**
     * Set a callback for the onScannedBot event.
     *
     * @param callback the scanned bot event callback
     * @return this builder
     */
    public TestBotBuilder onScannedBot(Consumer<ScannedBotEvent> callback) {
        this.onScannedBotCallback = callback;
        return this;
    }

    /**
     * Set a callback for the onHitBot event.
     *
     * @param callback the hit bot event callback
     * @return this builder
     */
    public TestBotBuilder onHitBot(Consumer<HitBotEvent> callback) {
        this.onHitBotCallback = callback;
        return this;
    }

    /**
     * Set a callback for the onHitWall event.
     *
     * @param callback the hit wall event callback
     * @return this builder
     */
    public TestBotBuilder onHitWall(Consumer<HitWallEvent> callback) {
        this.onHitWallCallback = callback;
        return this;
    }

    /**
     * Set a callback for the onBulletFired event.
     *
     * @param callback the bullet fired event callback
     * @return this builder
     */
    public TestBotBuilder onBulletFired(Consumer<BulletFiredEvent> callback) {
        this.onBulletFiredCallback = callback;
        return this;
    }

    /**
     * Set a callback for the onHitByBullet event.
     *
     * @param callback the hit by bullet event callback
     * @return this builder
     */
    public TestBotBuilder onHitByBullet(Consumer<HitByBulletEvent> callback) {
        this.onHitByBulletCallback = callback;
        return this;
    }

    /**
     * Set a callback for the onBulletHitBot event.
     *
     * @param callback the bullet hit bot event callback
     * @return this builder
     */
    public TestBotBuilder onBulletHitBot(Consumer<BulletHitBotEvent> callback) {
        this.onBulletHitBotCallback = callback;
        return this;
    }

    /**
     * Set a callback for the onDeath event.
     *
     * @param callback the death event callback
     * @return this builder
     */
    public TestBotBuilder onDeath(Consumer<DeathEvent> callback) {
        this.onDeathCallback = callback;
        return this;
    }

    /**
     * Set a callback for the onRoundStarted event.
     *
     * @param callback the round started event callback
     * @return this builder
     */
    public TestBotBuilder onRoundStarted(Consumer<RoundStartedEvent> callback) {
        this.onRoundStartedCallback = callback;
        return this;
    }

    /**
     * Set a callback for the onRoundEnded event.
     *
     * @param callback the round ended event callback
     * @return this builder
     */
    public TestBotBuilder onRoundEnded(Consumer<RoundEndedEvent> callback) {
        this.onRoundEndedCallback = callback;
        return this;
    }

    /**
     * Set a callback for the onGameStarted event.
     *
     * @param callback the game started event callback
     * @return this builder
     */
    public TestBotBuilder onGameStarted(Consumer<GameStartedEvent> callback) {
        this.onGameStartedCallback = callback;
        return this;
    }

    /**
     * Set a callback for the onGameEnded event.
     *
     * @param callback the game ended event callback
     * @return this builder
     */
    public TestBotBuilder onGameEnded(Consumer<GameEndedEvent> callback) {
        this.onGameEndedCallback = callback;
        return this;
    }

    /**
     * Set a callback for the run() method.
     * This is called once when the bot starts running.
     *
     * @param callback the run callback
     * @return this builder
     */
    public TestBotBuilder onRun(Runnable callback) {
        this.onRunCallback = callback;
        return this;
    }

    /**
     * Build the test bot with the configured settings.
     *
     * @param serverUrl the server URL to connect to
     * @return the configured test bot
     */
    public Bot build(URI serverUrl) {
        BotInfo.IBuilder botInfoBuilder = BotInfo.builder()
                .setName(name)
                .setVersion(version);

        for (String author : authors) {
            botInfoBuilder.addAuthor(author);
        }

        botInfoBuilder
                .setDescription(description)
                .setPlatform(platform)
                .setProgrammingLang(programmingLang);

        BotInfo botInfo = botInfoBuilder.build();

        return new ConfigurableBot(botInfo, serverUrl, this);
    }

    /**
     * Build the test bot using the default MockedServer URL.
     *
     * @return the configured test bot
     */
    public Bot build() {
        return build(MockedServer.getServerUrl());
    }

    /**
     * Internal bot class that delegates to callbacks.
     */
    private static class ConfigurableBot extends Bot {
        private final TestBotBuilder config;

        ConfigurableBot(BotInfo botInfo, URI serverUrl, TestBotBuilder config) {
            super(botInfo, serverUrl);
            this.config = config;
        }

        @Override
        public void run() {
            // Execute custom run callback if provided
            if (config.onRunCallback != null) {
                config.onRunCallback.run();
            }

            // Main loop based on behavior
            while (isRunning()) {
                switch (config.behavior) {
                    case PASSIVE:
                        // Do nothing, just go
                        break;
                    case AGGRESSIVE:
                        setForward(100);
                        setFire(1);
                        break;
                    case SCANNING:
                        setRadarTurnRate(45);
                        break;
                    case CUSTOM:
                        // Custom behavior is handled entirely by callbacks
                        break;
                }
                go();
            }
        }

        @Override
        public void onTick(TickEvent tickEvent) {
            // Apply behavior-specific tick actions
            switch (config.behavior) {
                case AGGRESSIVE:
                    // Continue firing if gun is cool
                    if (getGunHeat() == 0) {
                        setFire(1);
                    }
                    break;
                case SCANNING:
                    // Keep radar spinning
                    setRadarTurnRate(45);
                    break;
                default:
                    break;
            }

            // Invoke custom callback
            if (config.onTickCallback != null) {
                config.onTickCallback.accept(tickEvent);
            }
        }

        @Override
        public void onScannedBot(ScannedBotEvent scannedBotEvent) {
            // Behavior-specific actions
            if (config.behavior == BotBehavior.AGGRESSIVE) {
                // Turn towards scanned bot and fire
                double bearing = directionTo(scannedBotEvent.getX(), scannedBotEvent.getY());
                setTurnRate(bearing - getDirection());
                setFire(3);
            }

            if (config.onScannedBotCallback != null) {
                config.onScannedBotCallback.accept(scannedBotEvent);
            }
        }

        @Override
        public void onHitBot(HitBotEvent botHitBotEvent) {
            if (config.onHitBotCallback != null) {
                config.onHitBotCallback.accept(botHitBotEvent);
            }
        }

        @Override
        public void onHitWall(HitWallEvent botHitWallEvent) {
            // Behavior-specific actions
            if (config.behavior == BotBehavior.AGGRESSIVE) {
                // Reverse direction when hitting wall
                setForward(-100);
            }

            if (config.onHitWallCallback != null) {
                config.onHitWallCallback.accept(botHitWallEvent);
            }
        }

        @Override
        public void onBulletFired(BulletFiredEvent bulletFiredEvent) {
            if (config.onBulletFiredCallback != null) {
                config.onBulletFiredCallback.accept(bulletFiredEvent);
            }
        }

        @Override
        public void onHitByBullet(HitByBulletEvent hitByBulletEvent) {
            if (config.onHitByBulletCallback != null) {
                config.onHitByBulletCallback.accept(hitByBulletEvent);
            }
        }

        @Override
        public void onBulletHit(BulletHitBotEvent bulletHitBotEvent) {
            if (config.onBulletHitBotCallback != null) {
                config.onBulletHitBotCallback.accept(bulletHitBotEvent);
            }
        }

        @Override
        public void onDeath(DeathEvent deathEvent) {
            if (config.onDeathCallback != null) {
                config.onDeathCallback.accept(deathEvent);
            }
        }

        @Override
        public void onRoundStarted(RoundStartedEvent roundStartedEvent) {
            if (config.onRoundStartedCallback != null) {
                config.onRoundStartedCallback.accept(roundStartedEvent);
            }
        }

        @Override
        public void onRoundEnded(RoundEndedEvent roundEndedEvent) {
            if (config.onRoundEndedCallback != null) {
                config.onRoundEndedCallback.accept(roundEndedEvent);
            }
        }

        @Override
        public void onGameStarted(GameStartedEvent gameStartedEvent) {
            if (config.onGameStartedCallback != null) {
                config.onGameStartedCallback.accept(gameStartedEvent);
            }
        }

        @Override
        public void onGameEnded(GameEndedEvent gameEndedEvent) {
            if (config.onGameEndedCallback != null) {
                config.onGameEndedCallback.accept(gameEndedEvent);
            }
        }
    }
}

File: bot-api/dotnet/test/src/test_utils/TestBotBuilder.cs
using System;
using Robocode.TankRoyale.BotApi;
using Robocode.TankRoyale.BotApi.Events;

namespace Robocode.TankRoyale.BotApi.Tests.Test_utils;

/// <summary>
/// Builder/factory for creating configurable test bots.
///
/// This builder allows tests to create bots with specific behaviors without
/// creating separate bot classes for each test scenario. It supports:
/// <list type="bullet">
///   <item>Predefined behavior profiles (passive, aggressive, scanning)</item>
///   <item>Custom event handler callbacks for targeted testing</item>
///   <item>Fluent API for easy configuration</item>
/// </list>
///
/// Example usage:
/// <code>
/// var bot = TestBotBuilder.Create()
///     .WithBehavior(BotBehavior.Passive)
///     .OnTick(e => bot.TurnRate = 5)
///     .Build(serverUrl);
/// </code>
/// </summary>
public class TestBotBuilder
{
    /// <summary>
    /// Predefined bot behavior profiles.
    /// </summary>
    public enum BotBehavior
    {
        /// <summary>Does nothing - just calls Go() each tick.</summary>
        Passive,

        /// <summary>Moves forward and fires continuously.</summary>
        Aggressive,

        /// <summary>Continuously rotates radar to scan for enemies.</summary>
        Scanning,

        /// <summary>Custom behavior defined by callbacks only.</summary>
        Custom
    }

    // Bot info configuration
    private string _name = "TestBot";
    private string _version = "1.0";
    private string[] _authors = { "Test Author" };
    private string _description = "Test bot for unit tests";
    private string _platform = ".NET 8";
    private string _programmingLang = "C#";

    // Behavior configuration
    private BotBehavior _behavior = BotBehavior.Passive;

    // Event callbacks
    private Action<TickEvent>? _onTickCallback;
    private Action<ScannedBotEvent>? _onScannedBotCallback;
    private Action<HitBotEvent>? _onHitBotCallback;
    private Action<HitWallEvent>? _onHitWallCallback;
    private Action<BulletFiredEvent>? _onBulletFiredCallback;
    private Action<HitByBulletEvent>? _onHitByBulletCallback;
    private Action<BulletHitBotEvent>? _onBulletHitBotCallback;
    private Action<DeathEvent>? _onDeathCallback;
    private Action<RoundStartedEvent>? _onRoundStartedCallback;
    private Action<RoundEndedEvent>? _onRoundEndedCallback;
    private Action<GameStartedEvent>? _onGameStartedCallback;
    private Action<GameEndedEvent>? _onGameEndedCallback;
    private Action? _onRunCallback;

    private TestBotBuilder() { }

    /// <summary>
    /// Create a new TestBotBuilder instance.
    /// </summary>
    /// <returns>A new builder.</returns>
    public static TestBotBuilder Create() => new();

    /// <summary>
    /// Set the bot name.
    /// </summary>
    /// <param name="name">The bot name.</param>
    /// <returns>This builder.</returns>
    public TestBotBuilder WithName(string name)
    {
        _name = name;
        return this;
    }

    /// <summary>
    /// Set the bot version.
    /// </summary>
    /// <param name="version">The bot version.</param>
    /// <returns>This builder.</returns>
    public TestBotBuilder WithVersion(string version)
    {
        _version = version;
        return this;
    }

    /// <summary>
    /// Set the bot authors.
    /// </summary>
    /// <param name="authors">The bot authors.</param>
    /// <returns>This builder.</returns>
    public TestBotBuilder WithAuthors(params string[] authors)
    {
        _authors = authors;
        return this;
    }

    /// <summary>
    /// Set the bot description.
    /// </summary>
    /// <param name="description">The bot description.</param>
    /// <returns>This builder.</returns>
    public TestBotBuilder WithDescription(string description)
    {
        _description = description;
        return this;
    }

    /// <summary>
    /// Set the bot behavior profile.
    /// </summary>
    /// <param name="behavior">The behavior profile.</param>
    /// <returns>This builder.</returns>
    public TestBotBuilder WithBehavior(BotBehavior behavior)
    {
        _behavior = behavior;
        return this;
    }

    /// <summary>
    /// Set the onTick callback.
    /// </summary>
    /// <param name="callback">The callback to invoke on tick events.</param>
    /// <returns>This builder.</returns>
    public TestBotBuilder OnTick(Action<TickEvent> callback)
    {
        _onTickCallback = callback;
        return this;
    }

    /// <summary>
    /// Set the onScannedBot callback.
    /// </summary>
    /// <param name="callback">The callback to invoke when a bot is scanned.</param>
    /// <returns>This builder.</returns>
    public TestBotBuilder OnScannedBot(Action<ScannedBotEvent> callback)
    {
        _onScannedBotCallback = callback;
        return this;
    }

    /// <summary>
    /// Set the onHitBot callback.
    /// </summary>
    /// <param name="callback">The callback to invoke when hitting another bot.</param>
    /// <returns>This builder.</returns>
    public TestBotBuilder OnHitBot(Action<HitBotEvent> callback)
    {
        _onHitBotCallback = callback;
        return this;
    }

    /// <summary>
    /// Set the onHitWall callback.
    /// </summary>
    /// <param name="callback">The callback to invoke when hitting a wall.</param>
    /// <returns>This builder.</returns>
    public TestBotBuilder OnHitWall(Action<HitWallEvent> callback)
    {
        _onHitWallCallback = callback;
        return this;
    }

    /// <summary>
    /// Set the onBulletFired callback.
    /// </summary>
    /// <param name="callback">The callback to invoke when a bullet is fired.</param>
    /// <returns>This builder.</returns>
    public TestBotBuilder OnBulletFired(Action<BulletFiredEvent> callback)
    {
        _onBulletFiredCallback = callback;
        return this;
    }

    /// <summary>
    /// Set the onHitByBullet callback.
    /// </summary>
    /// <param name="callback">The callback to invoke when hit by a bullet.</param>
    /// <returns>This builder.</returns>
    public TestBotBuilder OnHitByBullet(Action<HitByBulletEvent> callback)
    {
        _onHitByBulletCallback = callback;
        return this;
    }

    /// <summary>
    /// Set the onBulletHit callback.
    /// </summary>
    /// <param name="callback">The callback to invoke when the bot's bullet hits another bot.</param>
    /// <returns>This builder.</returns>
    public TestBotBuilder OnBulletHit(Action<BulletHitBotEvent> callback)
    {
        _onBulletHitBotCallback = callback;
        return this;
    }

    /// <summary>
    /// Set the onDeath callback.
    /// </summary>
    /// <param name="callback">The callback to invoke when the bot dies.</param>
    /// <returns>This builder.</returns>
    public TestBotBuilder OnDeath(Action<DeathEvent> callback)
    {
        _onDeathCallback = callback;
        return this;
    }

    /// <summary>
    /// Set the onRoundStarted callback.
    /// </summary>
    /// <param name="callback">The callback to invoke when a round starts.</param>
    /// <returns>This builder.</returns>
    public TestBotBuilder OnRoundStarted(Action<RoundStartedEvent> callback)
    {
        _onRoundStartedCallback = callback;
        return this;
    }

    /// <summary>
    /// Set the onRoundEnded callback.
    /// </summary>
    /// <param name="callback">The callback to invoke when a round ends.</param>
    /// <returns>This builder.</returns>
    public TestBotBuilder OnRoundEnded(Action<RoundEndedEvent> callback)
    {
        _onRoundEndedCallback = callback;
        return this;
    }

    /// <summary>
    /// Set the onGameStarted callback.
    /// </summary>
    /// <param name="callback">The callback to invoke when the game starts.</param>
    /// <returns>This builder.</returns>
    public TestBotBuilder OnGameStarted(Action<GameStartedEvent> callback)
    {
        _onGameStartedCallback = callback;
        return this;
    }

    /// <summary>
    /// Set the onGameEnded callback.
    /// </summary>
    /// <param name="callback">The callback to invoke when the game ends.</param>
    /// <returns>This builder.</returns>
    public TestBotBuilder OnGameEnded(Action<GameEndedEvent> callback)
    {
        _onGameEndedCallback = callback;
        return this;
    }

    /// <summary>
    /// Set the onRun callback.
    /// </summary>
    /// <param name="callback">The callback to invoke in the Run() method.</param>
    /// <returns>This builder.</returns>
    public TestBotBuilder OnRun(Action callback)
    {
        _onRunCallback = callback;
        return this;
    }

    /// <summary>
    /// Build a bot using the default server URL (MockedServer.ServerUrl).
    /// </summary>
    /// <returns>A configured bot instance.</returns>
    public Bot Build()
    {
        return Build(MockedServer.ServerUrl);
    }

    /// <summary>
    /// Build a bot using the specified server URL.
    /// </summary>
    /// <param name="serverUrl">The server URL to connect to.</param>
    /// <returns>A configured bot instance.</returns>
    public Bot Build(Uri serverUrl)
    {
        var botInfoBuilder = BotInfo.Builder()
            .SetName(_name)
            .SetVersion(_version);

        foreach (var author in _authors)
        {
            botInfoBuilder.AddAuthor(author);
        }

        if (!string.IsNullOrEmpty(_description))
        {
            botInfoBuilder.SetDescription(_description);
        }

        botInfoBuilder.SetPlatform(_platform);
        botInfoBuilder.SetProgrammingLang(_programmingLang);

        var botInfo = botInfoBuilder.Build();

        var config = new TestBotConfig
        {
            Behavior = _behavior,
            OnTickCallback = _onTickCallback,
            OnScannedBotCallback = _onScannedBotCallback,
            OnHitBotCallback = _onHitBotCallback,
            OnHitWallCallback = _onHitWallCallback,
            OnBulletFiredCallback = _onBulletFiredCallback,
            OnHitByBulletCallback = _onHitByBulletCallback,
            OnBulletHitBotCallback = _onBulletHitBotCallback,
            OnDeathCallback = _onDeathCallback,
            OnRoundStartedCallback = _onRoundStartedCallback,
            OnRoundEndedCallback = _onRoundEndedCallback,
            OnGameStartedCallback = _onGameStartedCallback,
            OnGameEndedCallback = _onGameEndedCallback,
            OnRunCallback = _onRunCallback
        };

        return new ConfigurableTestBot(botInfo, serverUrl, config);
    }

    /// <summary>
    /// Internal configuration for the test bot.
    /// </summary>
    private class TestBotConfig
    {
        public BotBehavior Behavior { get; set; }
        public Action<TickEvent>? OnTickCallback { get; set; }
        public Action<ScannedBotEvent>? OnScannedBotCallback { get; set; }
        public Action<HitBotEvent>? OnHitBotCallback { get; set; }
        public Action<HitWallEvent>? OnHitWallCallback { get; set; }
        public Action<BulletFiredEvent>? OnBulletFiredCallback { get; set; }
        public Action<HitByBulletEvent>? OnHitByBulletCallback { get; set; }
        public Action<BulletHitBotEvent>? OnBulletHitBotCallback { get; set; }
        public Action<DeathEvent>? OnDeathCallback { get; set; }
        public Action<RoundStartedEvent>? OnRoundStartedCallback { get; set; }
        public Action<RoundEndedEvent>? OnRoundEndedCallback { get; set; }
        public Action<GameStartedEvent>? OnGameStartedCallback { get; set; }
        public Action<GameEndedEvent>? OnGameEndedCallback { get; set; }
        public Action? OnRunCallback { get; set; }
    }

    /// <summary>
    /// Internal bot implementation that delegates to callbacks.
    /// </summary>
    private class ConfigurableTestBot : Bot
    {
        private readonly TestBotConfig _config;

        public ConfigurableTestBot(BotInfo botInfo, Uri serverUrl, TestBotConfig config)
            : base(botInfo, serverUrl)
        {
            _config = config;
        }

        public override void Run()
        {
            // Call custom run callback if provided
            _config.OnRunCallback?.Invoke();

            // Behavior-specific run logic
            while (IsRunning)
            {
                switch (_config.Behavior)
                {
                    case BotBehavior.Passive:
                        // Do nothing, just wait for next turn
                        break;

                    case BotBehavior.Aggressive:
                        // Move forward and fire
                        SetForward(100);
                        SetFire(1);
                        break;

                    case BotBehavior.Scanning:
                        // Spin radar continuously
                        RadarTurnRate = Constants.MaxRadarTurnRate;
                        break;

                    case BotBehavior.Custom:
                        // No default behavior, rely on callbacks
                        break;
                }

                Go();
            }
        }

        public override void OnTick(TickEvent tickEvent)
        {
            _config.OnTickCallback?.Invoke(tickEvent);
        }

        public override void OnScannedBot(ScannedBotEvent scannedBotEvent)
        {
            // Behavior-specific actions
            if (_config.Behavior == BotBehavior.Aggressive)
            {
                // Turn gun toward scanned bot and fire
                var bearingToTarget = GunBearingTo(scannedBotEvent.X, scannedBotEvent.Y);
                SetTurnGunLeft(bearingToTarget);
                SetFire(2);
            }

            _config.OnScannedBotCallback?.Invoke(scannedBotEvent);
        }

        public override void OnHitBot(HitBotEvent botHitBotEvent)
        {
            // Behavior-specific actions
            if (_config.Behavior == BotBehavior.Aggressive)
            {
                // Fire at point-blank range
                SetFire(3);
            }

            _config.OnHitBotCallback?.Invoke(botHitBotEvent);
        }

        public override void OnHitWall(HitWallEvent botHitWallEvent)
        {
            // Behavior-specific actions
            if (_config.Behavior == BotBehavior.Aggressive)
            {
                // Reverse direction when hitting wall
                SetForward(-100);
            }

            _config.OnHitWallCallback?.Invoke(botHitWallEvent);
        }

        public override void OnBulletFired(BulletFiredEvent bulletFiredEvent)
        {
            _config.OnBulletFiredCallback?.Invoke(bulletFiredEvent);
        }

        public override void OnHitByBullet(HitByBulletEvent hitByBulletEvent)
        {
            _config.OnHitByBulletCallback?.Invoke(hitByBulletEvent);
        }

        public override void OnBulletHit(BulletHitBotEvent bulletHitBotEvent)
        {
            _config.OnBulletHitBotCallback?.Invoke(bulletHitBotEvent);
        }

        public override void OnDeath(DeathEvent deathEvent)
        {
            _config.OnDeathCallback?.Invoke(deathEvent);
        }

        public override void OnRoundStarted(RoundStartedEvent roundStartedEvent)
        {
            _config.OnRoundStartedCallback?.Invoke(roundStartedEvent);
        }

        public override void OnRoundEnded(RoundEndedEvent roundEndedEvent)
        {
            _config.OnRoundEndedCallback?.Invoke(roundEndedEvent);
        }

        public override void OnGameStarted(GameStartedEvent gameStartedEvent)
        {
            _config.OnGameStartedCallback?.Invoke(gameStartedEvent);
        }

        public override void OnGameEnded(GameEndedEvent gameEndedEvent)
        {
            _config.OnGameEndedCallback?.Invoke(gameEndedEvent);
        }
    }
}
